File: languages/go/04-expert/modularization-and-build-structure/example/pricing.go
package main

type lineItem struct {
	name      string
	quantity  int
	unitPrice float64
}

type invoiceSummary struct {
	subtotal      float64
	discountValue float64
	taxValue      float64
	total         float64
}

func buildSummary(items []lineItem, discountPercent, taxPercent float64) invoiceSummary {
	subtotal := 0.0
	for _, item := range items {
		subtotal += float64(item.quantity) * item.unitPrice
	}

	discountValue := subtotal * (discountPercent / 100.0)
	taxedBase := subtotal - discountValue
	taxValue := taxedBase * (taxPercent / 100.0)

	return invoiceSummary{
		subtotal:      subtotal,
		discountValue: discountValue,
		taxValue:      taxValue,
		total:         taxedBase + taxValue,
	}
}

File: languages/typescript/04-expert/modularization-and-build-structure/example/pricing.ts
// This helper example focuses on keeping pricing logic separate from orchestration and output code.

export type LineItem = {
    name: string;
    quantity: number;
    unitPrice: number;
};

export type InvoiceSummary = {
    subtotal: number;
    discountValue: number;
    taxValue: number;
    total: number;
};

export function buildSummary(
    items: LineItem[],
    discountPercent: number,
    taxPercent: number,
): InvoiceSummary {
    const subtotal = items.reduce(
        (sum, item) => sum + item.quantity * item.unitPrice,
        0,
    );
    const discountValue = subtotal * (discountPercent / 100);
    const taxedBase = subtotal - discountValue;
    const taxValue = taxedBase * (taxPercent / 100);

    return {
        subtotal,
        discountValue,
        taxValue,
        total: taxedBase + taxValue,
    };
}
